src/software/ai/navigator/path_planner/hrvo/hrvo_agent.cpp

```diff
@@ -571,60 +630,68 @@ std::optional<int> HRVOAgent::findIntersectingVelocityObstacle(
 
 Vector HRVOAgent::computePreferredVelocity(Duration time_step)
 {
-    double pref_speed                 = max_speed * PREF_SPEED_SCALE;
-    const auto current_path_point_opt = path.getCurrentPathPoint();
-
-    if (!current_path_point_opt.has_value() || pref_speed <= 0.01 || max_accel <= 0.01)
+    auto path_point_opt = path.getCurrentPathPoint();
+    if (!path_point_opt.has_value())
     {
-        // Used to avoid edge cases with division by zero
-        return Vector(0.0, 0.0);
+        return Vector();
     }
 
-    Point goal_position  = current_path_point_opt.value().getPosition();
-    double speed_at_goal = current_path_point_opt.value().getSpeed();
+    Point destination  = path_point_opt.value().getPosition();
+    Vector local_error = globalToLocalVelocity(destination - position, orientation);
 
-    Vector dist_vector_to_goal = goal_position - position;
-    auto dist_to_goal          = static_cast<float>(dist_vector_to_goal.length());
-
-    // d = (Vf^2 - Vi^2) / 2a
-    double start_linear_deceleration_distance =
-        std::abs((std::pow(speed_at_goal, 2) - std::pow(pref_speed, 2)) /
-                 (2 * max_accel)) *
-        DECEL_DIST_MULTIPLIER;
+    if (distance(destination, prev_dynamic_kp_destination) >
+        MAX_DESTINATION_CHANGE_THRESHOLD)
+    {
+        // Destination has significantly changed, recalculate dynamic kp.
+        // Relationship between initial distance to destination and dynamic kp was
+        // determined experimentally to have the robot decelerate as late as possible,
+        // without overshooting. The minimum distance which this function was tested on
+        // was 0.25m. More detail about the tests can be found on Notion from Apr 28,
+        // 2023.
+        double distance_for_kp      = std::max(0.25, local_error.length());
+        kp                          = 2.3 / (distance_for_kp + 0.4) + 1.5;
```

Comment:
what do you think about moving these class constants?

Reply:
Personally, I think it would just make it harder to read if each constant is moved to the header. The values are determined experimentally by just eye-balling a line of best fit, and I'm not sure if giving them a name would make them any more descriptive 🤷‍♂️